File: algorithms/invert_binary_tree.js

class BinaryTree {

    constructor() {
        this.root = null;
    }

    add(data) {
        const newNode = new Node(data)
        if (this.root === null) {
            this.root = newNode;
        } else {
            this.#addNode(this.root, newNode);
        }
    }

    #addNode(vertex, newNode) {
        if (newNode.data < vertex.data) {
            if (!vertex.left) {
                vertex.left = newNode;
            } else {
                this.#addNode(vertex.left, newNode);
            }
        } else {
            if (!vertex.right) {
                vertex.right = newNode;
            } else {
                this.#addNode(vertex.right, newNode);
            }
        }
    }
}

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

const tree = new BinaryTree();
tree.add(6);
tree.add(3);
tree.add(5);
tree.add(7);
tree.add(9);

console.log(tree);

const invert_binary_tree = (tree = {}) => {
    if (!tree.root) { return tree; }

    function invert_nodes(vertex) {
        if (!vertex.left && !vertex.right) { return vertex; }
        const tmp = vertex.left;
        vertex.left = vertex.right;
        vertex.right = tmp;
    }

    return invert_nodes(tree.root);
}

invert_binary_tree(tree);

console.log(tree)

File: algo_expert/by_topic/binary_trees/easy/InvertBinaryTree.java
public class InvertBinaryTree {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        TreeNode n3 = new TreeNode(3);
        TreeNode n5 = new TreeNode(5);
        TreeNode n1 = new TreeNode(1);
        root.left = n1;
        root.right = n5;
        n3.left = n1;
        invertTree(root);
    }

    public static TreeNode invertTree(TreeNode root) {
        invert(root);
        return root;
    }

    private static void invert(TreeNode node) {
        if (node == null) return;
        TreeNode temp = node.left;
        node.left = node.right;
        node.right = temp;

        if (node.left != null) {
            invert(node.left);
        }

        if (node.right != null) {
            invert(node.right);
        }
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
